[DOCS] DOC-363 how to configure a pandas datasource

Rachel-Reverie opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- New document on configuring a Pandas Datasource with examples for an InferredAssetDataConnector, ConfiguredAssetDataConnector, and RuntimeDataConnector as well as regex patterns for Data Assets that can only include on Batch and Data Assets that support returning multiple Batches.


After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-


### Definition of Done
Please delete options that are not relevant.

- [ ] I have made corresponding changes to the documentation


Thank you for submitting!

Rachel-Reverie committed:
DOC-363
- Rough draft: How to configure a Pandas Datasource: InferredAssetFilesystem Data Connector, single Batch regex
- Partial rough draft: How to configure a Pandas Datasource: InferredAssetFilesystem Data Connector, multi-Batch regex

Rachel-Reverie committed:
DOC-363
- completed rough draft: How to configure a Pandas Datasource: InferredAssetFilesystem Data Connector, multi-Batch regex

Rachel-Reverie committed:
Merge branch 'develop' into DOCS/DOC-363/how_to_configure_a_pandas_datasource

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5779&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5779&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/5779/28813127/0095c64c4dbec8463593dd55414c3b07118f2c52.svg)](https://app.codesee.io/r/reviews?pr=5779&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Rachel-Reverie committed:
DOC-363
- Corrected link

Rachel-Reverie committed:
DOC-363
- cleaned up Next Steps section

Rachel-Reverie committed:
DOC-363
- Correction to ToC link

Rachel-Reverie committed:
DOC-363
- typo in tip admonition block

Rachel-Reverie committed:
Merge branch 'develop' into DOCS/DOC-363/how_to_configure_a_pandas_datasource

Rachel-Reverie committed:
DOC-363
- Clarifications in steps 1-3

Rachel-Reverie committed:
DOC-363
- Added labels to code blocks

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
````diff
@@ -0,0 +1,355 @@
+```
+The `great_expectations` module will give you access to your Data Context, which is the entry point for working with a Great Expectations project.
+
+The `yaml` module from `ruamel` will be used in validating your Datasource's configuration.  Great Expectations will use a Python dictionary representation of your Datasource configuration when you add your Datasource to your Data Context.  However, Great Expectations saves configurations as `yaml` files, so when you validate your configuration you will need to convert it from a Python dictionary ...
````
I think this is the first time this step has been laid out so clearly. Thank you

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
````diff
@@ -0,0 +1,355 @@
+At this point, the configuration for your Datasource is merely:
+
+```python
+datasource_config = {}
````
I think we do try to have type hinting for code nowadays : 
```suggestion
datasource_config: dict = {}
```

Rachel-Reverie replied:
Thanks for the catch; updating that in the next commit.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
```diff
@@ -0,0 +1,355 @@
+
+### 4. Specify the Datasource class and module
+
+The `class_name` and `module_name` for your Datasource will almost always indicate the `Datasource` class found at `great_expectations.datasource`.  You may replace this with a specialized subclass, or a custom class, but for almost all regular purposes these two default values will suffice.  For the purposes of this guide, add those two values to their corresponding keys.
```
Similar point here. Thank you for laying this out so clearly

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
````diff
@@ -0,0 +1,355 @@
+
+```python
+datasource_config = {
+    "name": "my_datasource_name",  # Preferably name it something relevant
````
I think the name may have changed between here and the previous steps.

Rachel-Reverie replied:
It did.  I've updated the earlier code snippets, though they'll be updated again when once I move the code snippets into a script under test after the drafts are done.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
```diff
@@ -0,0 +1,355 @@
+- `group_names`: This is a list of names that correspond to the groups you defined in `pattern`'s regular expression.
+
+The `pattern` in `default_regex` will be matched against the files in your `base_directory`, and everything that matches against the first group in your regex will become a batch in a Data Asset that possesses the name of the matching text.  Any files that have a matching string for the first group will become Batches in the same Data Asset.
+
```
"Batch" or batch? Are they different?

Rachel-Reverie replied:
Batch, I just missed the capitalization.  I have a script that I run after drafting to catch any words in our Technical Terms glossary that aren't title cased, heh.

Rachel-Reverie replied:
Fixes for this and one other instance ("batches" instead of "Batches") will be in the next commit.

Shinnnyshinshin replied:
Thanks

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
```diff
@@ -0,0 +1,355 @@
+
+:::caution 
+
+If the value of `datasource_config["name"]` corresponds to a Datasource that is already defined in your Data Context, then using the above command will overwrite the existing Datasource.
+
```
❤️

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/pandas_components/inferred_asset_data_connector.md:
````diff
@@ -0,0 +1,174 @@
+
+The simplest way to do this is to define a group that consists of the entire file name.
+
+For this example, lets assume we have the following files in our `data` directory:
+- `yellow_tripdata_sample_2020-01.csv`
+- `yellow_tripdata_sample_2020-02.csv`
+- `yellow_tripdata_sample_2020-03.csv`
+
+In this case you could define the `pattern` key as follows:
+
+```python
+"pattern": "(.*)\\.csv"
+```
````
YES! love it

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/pandas_components/inferred_asset_data_connector.md:
````diff
@@ -0,0 +1,174 @@
+```python
+    "group_names": ["data_asset_name"],
+```
+
+Looking back at our sample files, this regex will result in the `InferredAssetFilesystemDataConnector` providing three Data Assets, which can be accessed by the portion of the file that matches the first group in our regex.  In future workflows you will be able to refer to one of these Data Assets in a Batch Request py providing one of the following `data_asset_name`s:
+- `yellow_tripdata_sample_2020-01`
+- `yellow_tripdata_sample_2020-02`
+- `yellow_tripdata_sample_2020-03`
+
````
```suggestion

Looking back at our sample files, this regex will result in the `InferredAssetFilesystemDataConnector` providing three Data Assets, which can be accessed by the portion of the file that matches the first group in our regex.  In future workflows you will be able to refer to one of these Data Assets in a Batch Request by providing one of the following `data_asset_name`s:
- `yellow_tripdata_sample_2020-01`
- `yellow_tripdata_sample_2020-02`
- `yellow_tripdata_sample_2020-03`

```

Shinnnyshinshin replied:
changed "py" to "by"

Rachel-Reverie replied:
It took me a moment to see what was changed.  Good catch, thank you.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/pandas_components/inferred_asset_data_connector.md:
```diff
@@ -0,0 +1,174 @@
+
+For more information on the special characters and mechanics of matching and grouping strings with regular expressions, please see [the Python documentation on the `re` module](https://docs.python.org/3/library/re.html).
+
+:::
```
love it

Shinnnyshinshin commented on sidebars.js:
```diff
@@ -103,6 +103,7 @@ module.exports = {
             'guides/connecting_to_your_data/how_to_choose_which_dataconnector_to_use',
             'guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data',
             'guides/connecting_to_your_data/how_to_configure_an_inferredassetdataconnector',
+            'guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource',
```
I actually had some trouble finding the document in the netlify-rendered website.. Is this a new section under "connecting to your data?"

Rachel-Reverie replied:
The folder `datasource_configuration` is new, but currently its contents are being put in the core skills section under connecting to your data.  That's probably going to see some overhaul once all the docs are done, as some of the older ones will no longer be needed since they will be duplicated by the content in these new ones.

Shinnnyshinshin reviewed (approved):
LGTM @Rachel-Reverie. I think it's the clearest I've seen these concepts explained, ever. Thank you so much. 

 Some minor suggestions, and a question regarding where the doc is supposed to go

Rachel-Reverie committed:
DOC-363
- corrected title casing of technical terms
- corrected Datasource name in earlier snippets
- added type hints to code snippets
- removed titles on code blocks; visually they were extraneous since everything is Python.

Rachel-Reverie committed:
DOC-363
- consolidated some duplicated text into compnent .mdx files.
- adds the file that will contain instructions for defining assets in a Configured Data Connector.

Rachel-Reverie committed:
DOC-363
- consolidated some duplicated text into component .mdx files.
- content for the configured Data Connector portion of the document.

Rachel-Reverie committed:
DOC-363
- corrected relative path to data folder

Rachel-Reverie committed:
DOC-363
- Added optional step for adding multiple Data Connectors to the configuration.

Rachel-Reverie committed:
DOC-363
- Added `RuntimeDataConnector` configuration guidance.

Rachel-Reverie committed:
DOC-363
- fixed a typo remaining in code block code types

Rachel-Reverie committed:
DOC-363
- fixed a typo remaining in code block code types, in more files

Rachel-Reverie committed:
DOC-363
- polished and clarified RuntimeDataConnector guidance.

Rachel-Reverie committed:
DOC-363
- correct typo in the `data_connectors` key

Rachel-Reverie committed:
DOC-363
- minor corrections

rdodev commented on docs/guides/connecting_to_your_data/datasource_configuration/components/intro_for_single_or_multi_batch_configuration.mdx:
```diff
@@ -0,0 +1,7 @@
+This means that when configuring your Data Connector's regular expression, you have the option to design it so that the Data Connector is only capable of returning a single Batch per Data Asset, or so that it is capable of returning multiple Batches grouped into individual Data Assets.  Each type of configuration is useful in certain cases, so we will provide examples of both.
```
Minor comment: some thing about 'design' and RegEx doesn't quite fit :D maybe build or implement? Again not a big deal.

Rachel-Reverie replied:
I like "implement" better, too.  I'll change it to that.

Rachel-Reverie committed:
DOC-363
- update to better word choice

Rachel-Reverie committed:
DOC-363
- Added examples for `glob_directive` in Inferred Data Asset guidance.

Rachel-Reverie committed:
DOC-363
- cleaned up `glob_directive` example.

Commit:
Merge branch 'develop' into DOCS/DOC-363/how_to_configure_a_pandas_datasource

Shinnnyshinshin commented:
Had another look this morning with the updated changes, and everything looks great ✅ thank you @Rachel-Reverie

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/datasource_configuration/how_to_configure_a_pandas_datasource.md:
````diff
@@ -0,0 +1,546 @@
+    }
+}
+```
+:::info Optional parameter: `glob_directive`
+
+The `glob_directive` parameter is provided to give the `DataConnector` information about the directory structure to expect when identifying source data files to check against each Data Asset's `default_regex`.  If you do not specify a value for `glob_directive` a default value of `"*"` will be used. This will cause your Data Asset to check all files in the folder specified by `base_directory` to  ...
+
+Overriding the `glob_directive` by providing your own value will allow your Data Connector to traverse subdirectories or otherwise alter which source data files are compared against your Data Connector's `default_regex`.
+
+For example, assume your source data is in files contained by subdirectories of your `base_folder`, like so:
+- 2019/yellow_taxidata_2019_01.csv
+- 2020/yellow_taxidata_2020_01.csv
+- 2021/yellow_taxidata_2021_01.csv
+- 2022/yellow_taxidata_2022_01.csv
+
+To include all of these files, you would need to tell the Data connector to look for files that are nested one level deeper than the `base_directory` itself. 
+
+ You would do this by setting the `glob_directive` key in your Data Connector config to a value of `"*/*"`.  This value will cause the Data Connector to look for regex matches against the file names for all files found in any subfolder of your `base_directory`.  Such an entry would look like:
+
````
really really like this explanation. Thank you @Rachel-Reverie

Rachel-Reverie committed:
Merge branch 'develop' into DOCS/DOC-363/how_to_configure_a_pandas_datasource

anthonyburdi commented:
Merged bypassing codeowner check since @donaldheppner is OOO
